enonic/src/main/resources/site/parts/comments-list/comments-list.ts

```diff
@@ -0,0 +1,230 @@
+  readonly comments: ReadonlyArray<ContentTree>
+  readonly serviceUrl: string
+  readonly parentPath?: string
+  readonly pathSelf: string
+  readonly loginUrl?: string;
+  readonly logOut?: string
+  readonly isAuthenticated: boolean;
+}
+
+function createThymeleafParams(
+  userMapping: UserMapping,
+  comments: ReadonlyArray<Content<Comment>>,
+  isAuthenticated: boolean,
+  path: string,
+  partConfig: CommentsListPartConfig,
+  parentPath: string,
+): ThymeleafParams {
+  return {
+    comments: createCommentsTree(userMapping, partConfig.headingTag)(comments),
+    serviceUrl: serviceUrl('comment'),
+    parentPath: parentPath,
+    pathSelf: path,
+    loginUrl: getLoginUrl(partConfig.loginPageId),
+    logOut: logoutUrl({}),
+    isAuthenticated
+  }
+}
+function getLoginUrl(partLoginPageId?: string) {
+
+  return app.config.loginPageId ?
+    pageUrl(app.config.loginPageId) :
+    partLoginPageId ?
+    pageUrl(partLoginPageId) :
+    loginUrl({});
+}
+
+function notNullOrUndefined(str: string | undefined): boolean {
+  return (str !== undefined) && (str !== null);
+}
+
+export const byParentFirst = pipe(
+  ordBoolean,
+  contramap((c: Content<Comment>) => notNullOrUndefined(c.data.parentId))
+)
+
+export function childrenToParent(acc: ReadonlyArray<ContentTree>, curr: ContentTree): ReadonlyArray<ContentTree> {
+  if (curr.parentId) {
+    return acc.map((comment: ContentTree) => {
+      if(comment._id === curr.parentId) {
+        const children = comment.children ?? []
+        return {...comment, children: [...children, curr]} as ContentTree
+      }
+      return comment
+    })
+  } else {
+    return [...acc, curr]
+  }
+}
+
+export interface ContentTree {
+  _id: string
+  title: string
+  body: string
+  createdTime: string
+  createdTimeDisplay: string
+  parentId?: string,
+  creatorDisplayName?: string;
+  children?: ReadonlyArray<ContentTree>
+}
+
+type HeadingTag = "h2" | "h3" | "h4" | "h5"
+const commentResponseMap: Record<HeadingTag, string> = {
+  h2: "h3",
+  h3: "h4",
+  h4: "h5",
+  h5: "h6"
+}
+
+export function createCommentsTree(userMapping: UserMapping, headingTag: HeadingTag): (c: ReadonlyArray<Content<Comment>>) => ReadonlyArray<ContentTree> {
+  return flow(A.sort(byParentFirst), A.map(createCommentTree(headingTag, userMapping)), A.reduce([], childrenToParent))
+}
+function createCommentTree (headingTag: HeadingTag, userMapping: UserMapping) {
+  return function mapCommentToContentTree(comment: Content<Comment>) {
+    const heading = createTemplateTag(headingTag, comment.data)
+    return {
+      _id: comment._id,
+      heading:  processHtml({value: heading}),
+      title: comment.data.title,
+      body: comment.data.body,
+      creatorDisplayName: userMapping[comment.creator],
+      parentId: comment.data.parentId,
+      createdTime: comment.createdTime,
+      createdTimeDisplay: pipe(
+        comment.createdTime,
+        parseISO,
+        formatWithOptions({}, 'dd.MM.yyyy')
+      )
```

Comment:
I would pull the `pipe` out in its own separate function, to avoid too much nesting.